lib/WebSocket.js

```diff
@@ -17,6 +17,43 @@ wsAppHealth.on('connection', async function connection(ws, request) {
   ws.on('close', () => clearInterval(healthChecker));
 });
 
+const wsAppLogs = new WebSocket.Server({ noServer: true});
+wsAppLogs.on('connection', async function connection(ws, request) {
+  const appId = Number(url.parse(request.url, true).query.appId);
+  const app = await App.findByPk(appId);
+  const managerNode = await DockerWrapper.getManagerNodeInstance();
+  const service = managerNode.getService(`${app.title}_web`);
+  const duplex = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' });
+
+  const options = {
+    stdout: true,
+    stderr: true,
+    follow: true,
+  };
+
+  service.logs(options, function handler(err, stream) {
+    // I noticed when simply piping the stream to the WS client that weird
+    // artifacts would sometimes be included at the beginning of lines.
+    // See: https://github.com/moby/moby/issues/32794
+    //
+    // It turns out that Dockerode (or really, Docker Modem, which Dockerode uses
+    // internally) comes with a nice piece of functionality for handling this,
+    // called `demuxStream`
+    // More info here: https://github.com/apocas/dockerode/issues/259#issuecomment-242434163
+    service.modem.demuxStream(stream, duplex, duplex);
```

Comment:
So the stream is initially multiplex (with stdout and stderr), but you pass it to this function and it separates them into two separate streams (in this case you still send both through the WebsSocket)? Is it the case that the initial stream that you pass in is the unparsed JSON, but demuxed streams that you send over the socket are parsed?

Reply:
My mental model of how the Docker streams work under the hood are still in the process of maturing, but I think the way you've put it summarizes it.

[This comment](https://github.com/moby/moby/issues/32794#issuecomment-297151440) on the Moby project talks about the package that is used internally to actually multiplexes the logs into a single stream. It seems that in doing so some metadata about the streams themselves are intermixed into the result (in order to know what is what). The demuxer seems to know _how_ to separate these streams safely and then they can be sent over the socket.